### PR TITLE
Implementation of Ebay OAuth2 strategy for OmniAuth

### PR DESCRIPTION
**Pull request by IDolgirev**
Hi!
Even though the approach of this implementation is close to currently approved one(which IMO is definitely good) it slightly differs in configuration details and scope handling and I hope it deserves to be reviewed.

Thanks in advance!

### CODE REVIEW

File: lib/omniauth/ebay-oauth/user_info.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        @info = info['GetUserResponse']['User']
+      end
+
+      def uid
+        @info['EIASToken']
+      end
+
+      def info
+        {
+          username:   @info['UserID'],
```
Comment (Envek) [inline lib/omniauth/ebay-oauth/user_info.rb]: It's better to use `nickname` to conform with OmniAuth guidelines: https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema#schema-10-and-later

File: lib/omniauth/strategies/ebay.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      option :sandbox, false
+      option :client_options, auth_scheme: :basic_auth
+
+      uid   { raw_info.uid }
+      info  { raw_info.info }
+      extra { raw_info.extra }
+
+      def credentials
+        super.tap do |hash|
+          hash.merge!('refresh_token_expires_in' => access_token['refresh_token_expires_in'].to_i)
```
Comment (Envek) [inline lib/omniauth/strategies/ebay.rb]: `oauth2` gem actually converts access token lifetime duration from `expires_in` field to unix timestamp and exposes it in `expires_at` field (see [OAuth2::AccessToken#initialize](https://github.com/intridea/oauth2/blob/452a0853286fc0d2919037598ac891dd3e7b1d19/lib/oauth2/access_token.rb#L50)) and `omniauth-oauth2` [uses only this field](https://github.com/intridea/omniauth-oauth2/blob/464fcef6c0a5f1a7526b9701cec37f342ce57ae3/lib/omniauth/strategies/oauth2.rb#L42). I think it is better to do the same with refresh token for the sake of consistency.

Please change it to `refresh_token_expires_at` (returning Unix timestamp of token expiration).

File: lib/omniauth/ebay-oauth/api_scope_handler.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    class ApiScopeHandler # :nodoc:
+      def initialize(scopes)
+        @scopes = scopes || ''
+      end
+
+      def scopes_list
+        scopes = [DEFAULT_SCOPE]
+
+        unless @scopes.empty?
+          @scopes.split(SCOPE_DELIMITER).each do |scope|
```
Comment (Envek) [inline lib/omniauth/ebay-oauth/api_scope_handler.rb]: Why not just use arrays? Or use the `Array()` function here to let user to provide single value.

File: lib/omniauth/ebay-oauth/user_info.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+          first_name: name.split.first,
+          last_name:  name.split.last,
+          email:      @info['Email'],
+          phone:      @info['RegistrationAddress']['Phone'],
+          country:    @info['RegistrationAddress']['Country']
+        }
+      end
+
+      def extra
+        @info
```
Comment (Envek) [inline lib/omniauth/ebay-oauth/user_info.rb]: Same [OmniAuth guidelines](https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema#schema-10-and-later) prescribe, that all gathered raw info should be placed in `raw_info` key. Like this:

```json
{"extra":{"raw_info":{"AboutMePage":"false","EIASToken":"…"}}}
```

**Review by Envek — CHANGES_REQUESTED** [review]
Thank you very much for your contribution! Sure it deserves to be reviewed!
 
For the changes, required for smooth upgrading of existing apps see individual review comments.

The code is looking good to me in general, but module `UserInfoHelper` IMO is unneeded. I personally prefer to do `stub_request` once in `before` hook and all variable things for it (like response body or status code) extract to `let`s. And XMLs are best when they're ~dead~ in fixture files ;-)

But yes, it will be difficult to choose which PR to merge.

**Comment by IDolgirev** [comment]
Thank you very much for the review and comments!

I believe I've fixed all of them :)
Also I want to know your thoughts about assigning default public scope to token or append it to list of desired scopes. Leave it in current implementation as is?

**Comment by Envek** [comment]
I can't imagine why and when someone could need a token without an even public scope but who knows…

So how it is done in the current implementation is fine for me, at least for now.

But maybe it worth to have public scope to be the default but allow the user to get tokens without scopes at all by passing an empty array for example.

**Review by Envek — APPROVED** [review]

**Comment by Envek** [comment]
I'm closing this PR in favor of #2. I'm very grateful for your work and patience in resolving issues.

This wasn't an easy decision as all three submitted pull requests were fully functional solutions, differing in some subjective things.

This project is now alive and if you notice that it lacks some useful features, have bugs or needs polishing — pull requests are always welcome!

I learned something about OAuth while reviewing this PR (even after working with this protocol before) and hope that you have learned a lot too.

**Comment by IDolgirev** [comment]
I'm absolutely OK with that - at least #2 was submitted much faster :)
Thanks for your comments and good luck!

**Pull request closed without merging**